src/iocore/net/BIO_fastopen.cc

```diff
@@ -21,12 +21,117 @@
  *  limitations under the License.
  */
 
+#include <openssl/opensslv.h>
+
 #include "P_Net.h"
 #include "iocore/eventsystem/SocketManager.h"
 #include "tscore/ink_assert.h"
+#include "tscore/ink_config.h"
 
 #include "BIO_fastopen.h"
 
+#if defined(BORINGLIKE)
+#error
+#elif defined(OPENSSL_IS_BORINGSSL) || defined(OPENSSL_IS_AWSLC)
+#define BORINGLIKE 1
+#else
+#define BORINGLIKE 0
+#endif
+
+#if defined(HAVE_BIO_GET_EX_NEW_INDEX) && defined(HAVE_BIO_GET_EX_DATA) && defined(HAVE_BIO_SET_EX_DATA)
+
+// Thwart boringssl's thwarting of ex data for custom BIOs.  Boring fails to export these functions.
```

Comment:
I agree that BoringSSL's API is broken, but do we need to use these on code for BoringSSL?

Reply:
What alternative do you propose?

Reply:
Remove theses exports, and use empty callbacks for BoringSSL.

Reply:
I don't quite understand why OpenSSL says this though. BoringSSL doesn't say this.

> When setting exdata, the pointer specified with a particular index is saved, and returned on a subsequent "get" call. If the application is going to release the data, it must make sure to set a NULL value at the index, to avoid likely double-free crashes.

Reply:
I don't know why either.  In our case, the ex data is not a pointer to an allocated resource, so there is no risk of a double deallocation or a leak.  I only clear the ex data pointer to know it was free'd when looking at a coredump.